### PR TITLE
Fix GID allocator

### PR DESCRIPTION
**Pull request by RomanBednar**
**Is this a bug fix or adding new feature?**
Bug fix.

**What is this PR about? / Why do we need it?**

Dynamically provisioning EFS volume creates an access point (AP) for which it allocates a GID from default range or a range defined in SC parameters.

The GID allocator did not check what GIDs might be already in use. For example let's say we have an EFS with access point for `/mydir` path which is owned by GID `1000` with permissions of `770` that grant rwx access to group `1000`. Then we create another access point with same path or "higher" (can be also `/`) with the same GID of `1000`. Now whoever mounts the second access point will have rwx access to `/mydir` because the GID matches. The old code does not prevent this and so volume creation done by the driver might result in creating access point with a GID that is already used, and when it gets mounted to a pod this pod might get access to other pods data.

Additionally, when user changed the SC parameters it was not reflected in the driver (controller pod) and GID allocator kept allocating from the range that got in first (for a given FS ID). In order for GID range changes to take effect the pod had to be restarted.

**What testing is done?** 

- current unit tests
- new unit test added
- manual testing

### CODE REVIEW

**Comment by RomanBednar** [comment]
Here is what I've tested manually to verify the patches:

## Test GID allocator avoids used GIDS

1. Create one PVC dynamically with OCP and one directly in AWS cloud with GID of `6999999`.

```
$ aws efs describe-access-points --file-system-id fs-0c1bf51c4cc8a5771
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-0b38ad3470381476b    fsap-0b38ad3470381476b  pvc-a0e372b1-2f63-4561-a117-56df9df8d693        fs-0c1bf51c4cc8a5771    available                     269733383066
POSIXUSER       7000000 7000000
ROOTDIRECTORY   /dynamic_provisioning/pvc-a0e372b1-2f63-4561-a117-56df9df8d693
CREATIONINFO    7000000 7000000 700
TAGS    efs.csi.aws.com/cluster true
TAGS    kubernetes.io/cluster/rbednar-mycluster-01-l8wjr        owned
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-051deabd137c8ea32    fsap-051deabd137c8ea32  console-4d9cd57f-8192-4e14-8e2d-eb4144c8d5fb    fs-0c1bf51c4cc8a5771    available             test-ap 269733383066
POSIXUSER       6999999 6999999
SECONDARYGIDS   0
ROOTDIRECTORY   /
TAGS    Name    test-ap
```

2. Create a new PVC with dynamic provisioning with the same FS ID

3. Observe the used GID was detected and next available (`6999998`) got assigned.

```
$ aws efs describe-access-points --file-system-id fs-0c1bf51c4cc8a5771
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-0b38ad3470381476b    fsap-0b38ad3470381476b  pvc-a0e372b1-2f63-4561-a117-56df9df8d693        fs-0c1bf51c4cc8a5771    available                     269733383066
POSIXUSER       7000000 7000000
ROOTDIRECTORY   /dynamic_provisioning/pvc-a0e372b1-2f63-4561-a117-56df9df8d693
CREATIONINFO    7000000 7000000 700
TAGS    efs.csi.aws.com/cluster true
TAGS    kubernetes.io/cluster/rbednar-mycluster-01-l8wjr        owned
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-051deabd137c8ea32    fsap-051deabd137c8ea32  console-4d9cd57f-8192-4e14-8e2d-eb4144c8d5fb    fs-0c1bf51c4cc8a5771    available             test-ap 269733383066
POSIXUSER       6999999 6999999
SECONDARYGIDS   0
ROOTDIRECTORY   /
TAGS    Name    test-ap
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-05066babbad2648ac    fsap-05066babbad2648ac  pvc-1e332980-9c4f-41d4-9f03-aa47dd3feb37        fs-0c1bf51c4cc8a5771    available                     269733383066
POSIXUSER       6999998 6999998
ROOTDIRECTORY   /dynamic_provisioning/pvc-1e332980-9c4f-41d4-9f03-aa47dd3feb37
CREATIONINFO    6999998 6999998 700
TAGS    efs.csi.aws.com/cluster true
TAGS    kubernetes.io/cluster/rbednar-mycluster-01-l8wjr        owned
```

## Test that GID allocator can re-use released GIDs

4. Remove the extra AP (`test-ap`) that was created manually and create another PVC with dynamic provisioning. Now that the GID (`6999999`) is released we can see this GID is assigned to the new dynamically provisioned AP because it is now the next available GID.

```
$ aws efs describe-access-points --file-system-id fs-0c1bf51c4cc8a5771
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-0b38ad3470381476b    fsap-0b38ad3470381476b  pvc-a0e372b1-2f63-4561-a117-56df9df8d693        fs-0c1bf51c4cc8a5771    available             269733383066
POSIXUSER       7000000 7000000
ROOTDIRECTORY   /dynamic_provisioning/pvc-a0e372b1-2f63-4561-a117-56df9df8d693
CREATIONINFO    7000000 7000000 700
TAGS    efs.csi.aws.com/cluster true
TAGS    kubernetes.io/cluster/rbednar-mycluster-01-l8wjr        owned
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-05066babbad2648ac    fsap-05066babbad2648ac  pvc-1e332980-9c4f-41d4-9f03-aa47dd3feb37        fs-0c1bf51c4cc8a5771    available             269733383066
POSIXUSER       6999998 6999998
ROOTDIRECTORY   /dynamic_provisioning/pvc-1e332980-9c4f-41d4-9f03-aa47dd3feb37
CREATIONINFO    6999998 6999998 700
TAGS    efs.csi.aws.com/cluster true
TAGS    kubernetes.io/cluster/rbednar-mycluster-01-l8wjr        owned
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-0d5ffbeba4c86ef2d    fsap-0d5ffbeba4c86ef2d  pvc-a8329ba9-ca9c-4164-b4fb-41fcdaba6014        fs-0c1bf51c4cc8a5771    available             269733383066
POSIXUSER       6999999 6999999
ROOTDIRECTORY   /dynamic_provisioning/pvc-a8329ba9-ca9c-4164-b4fb-41fcdaba6014
CREATIONINFO    6999999 6999999 700
TAGS    efs.csi.aws.com/cluster true
TAGS    kubernetes.io/cluster/rbednar-mycluster-01-l8wjr        owned
```

## Test GID Range propagation

1. Recreate a storage class with `gidRangeStart` and  `gidRangeEnd` parameters

```
$ oc get sc/efs-sc -o yaml
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  creationTimestamp: "2022-12-08T10:48:42Z"
  name: efs-sc
  resourceVersion: "102207"
  uid: 6c66b91c-1aed-4b7e-8d4e-1518748b6e29
mountOptions:
- tls
parameters:
  basePath: /dynamic_provisioning
  directoryPerms: "700"
  fileSystemId: fs-0c1bf51c4cc8a5771
  gidRangeEnd: "2000"
  gidRangeStart: "1000"
  provisioningMode: efs-ap
provisioner: efs.csi.aws.com
reclaimPolicy: Delete
volumeBindingMode: Immediate
```

2. Without restarting any driver pods create another PVC and validate the new range was propagated correctly and new AP has GID `2000`

```
$ aws efs describe-access-points --file-system-id fs-0c1bf51c4cc8a5771
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-0b38ad3470381476b    fsap-0b38ad3470381476b  pvc-a0e372b1-2f63-4561-a117-56df9df8d693        fs-0c1bf51c4cc8a5771    available             269733383066
POSIXUSER       7000000 7000000
ROOTDIRECTORY   /dynamic_provisioning/pvc-a0e372b1-2f63-4561-a117-56df9df8d693
CREATIONINFO    7000000 7000000 700
TAGS    efs.csi.aws.com/cluster true
TAGS    kubernetes.io/cluster/rbednar-mycluster-01-l8wjr        owned
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-05066babbad2648ac    fsap-05066babbad2648ac  pvc-1e332980-9c4f-41d4-9f03-aa47dd3feb37        fs-0c1bf51c4cc8a5771    available             269733383066
POSIXUSER       6999998 6999998
ROOTDIRECTORY   /dynamic_provisioning/pvc-1e332980-9c4f-41d4-9f03-aa47dd3feb37
CREATIONINFO    6999998 6999998 700
TAGS    efs.csi.aws.com/cluster true
TAGS    kubernetes.io/cluster/rbednar-mycluster-01-l8wjr        owned
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-0d5ffbeba4c86ef2d    fsap-0d5ffbeba4c86ef2d  pvc-a8329ba9-ca9c-4164-b4fb-41fcdaba6014        fs-0c1bf51c4cc8a5771    available             269733383066
POSIXUSER       6999999 6999999
ROOTDIRECTORY   /dynamic_provisioning/pvc-a8329ba9-ca9c-4164-b4fb-41fcdaba6014
CREATIONINFO    6999999 6999999 700
TAGS    efs.csi.aws.com/cluster true
TAGS    kubernetes.io/cluster/rbednar-mycluster-01-l8wjr        owned
ACCESSPOINTS    arn:aws:elasticfilesystem:us-east-1:269733383066:access-point/fsap-04755a3ae8f8b7701    fsap-04755a3ae8f8b7701  pvc-fe36200c-bdb8-49ec-97bc-f9fb4e570627        fs-0c1bf51c4cc8a5771    available             269733383066
POSIXUSER       2000    2000
ROOTDIRECTORY   /dynamic_provisioning/pvc-fe36200c-bdb8-49ec-97bc-f9fb4e570627
CREATIONINFO    2000    2000    700
TAGS    efs.csi.aws.com/cluster true
TAGS    kubernetes.io/cluster/rbednar-mycluster-01-l8wjr        owned
```

**Comment by RomanBednar** [comment]
@wongma7 @nckturner what do you think about the change?

File: Dockerfile [hunk 1]
```diff
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-FROM golang:1.17 as builder
```
Comment (Ashley-wenyizha) [inline Dockerfile:15]: Could we separate Go upgrade in a separate PR?

**Comment by Ashley-wenyizha** [comment]
Hi Roman

Thanks for providing this fix. We are currently going through your code changes internally from EFS. 

One thing we want to call out is, also as you pointed out here https://github.com/kubernetes-sigs/aws-efs-csi-driver/issues/693#issuecomment-1357298780

that with the patch,  it would check all possible GIDs currently is 120 from given range each time a volume is created and this might not scale well as we have plan to increase this to 1000 early next year. AP increase internally dev is complete but waiting for deployment and official launch.

We will need to do a performance test upon the new limit and get back to you. Sorry about the potential delay, but what to let you know that we are looking into this and having internal tracking as well. Will provide more update upon this once we have more performance testing data.

Thank you!

File: pkg/driver/gid_allocator.go [hunk 2]
```diff
@@ -87,13 +60,66 @@ func (g *GidAllocator) removeFsId(fsId string) {
...
+	var availableGids []int
+
+	availableGids = make([]int, gidMax-gidMin+1)
+	if requestedRange > 120 {
+		klog.Warningf("Requested GID range (%v:%v) exceeds EFS Access Point limit (%v) per Filesystem. Driver will not allocate GIDs outside of this limit.", gidMin, gidMax, ACCESS_POINT_PER_FS_LIMIT)
+		availableGids = make([]int, ACCESS_POINT_PER_FS_LIMIT)
+	}
+
+	val := gidMax
+	for i := range availableGids {
```
Comment (mskanth972) [inline pkg/driver/gid_allocator.go:99]: Can we remove the loop and modify line no:106 as below  to improve the performance?
`for gid := gidMax; gid >= gidMin; gid--`

File: Dockerfile (hunk 1 above)
Comment (RomanBednar) [reply Dockerfile:15]: Sure, opened here: https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/867

File: pkg/driver/gid_allocator.go (hunk 2 above)
Comment (RomanBednar) [reply pkg/driver/gid_allocator.go]: Indeed, changed.

**Comment by RomanBednar** [comment]
Hello @Ashley-wenyizha, just checking in - anything new regarding the performance test?

**Comment by mskanth972** [comment]
Hi @RomanBednar, dynamically provisioned access points have a randomly generated folder names for example `/dynamic_provisioning/pvc-1e332980-9c4f-41d4-9f03-aa47dd3feb37`. Can you explain how can two different pods with different access points can mount the same folder?

File: pkg/driver/gid_allocator.go [hunk 3]
```diff
@@ -1,84 +1,57 @@
...
-func (h IntHeap) Len() int {
-	return len(h)
-}
-func (h IntHeap) Less(i, j int) bool {
-	return h[i] < h[j]
-}
-func (h IntHeap) Swap(i, j int) {
-	h[i], h[j] = h[j], h[i]
-}
+var ACCESS_POINT_PER_FS_LIMIT int = 120
```
Comment (jsafrane) [inline pkg/driver/gid_allocator.go:14]: This is now 1000: 
https://www.amazonaws.cn/en/new/2023/amazon-efs-increases-the-maximum-number-of-access-points-per-file-system/
https://docs.aws.amazon.com/efs/latest/ug/limits.html

**Comment by jsafrane** [comment]
I'd like to point out that this PR would also fix excessive memory consumption of the CSI driver when there is a large GID range configured - it will build a huge heap with all available GIDs, while only ~1000 can be ever used.

**Comment by RomanBednar** [comment]
/retest

CI verify job was failing due to old golang version which is now updated to 1.20

**Comment by RomanBednar** [comment]
@Ashley-wenyizha @mskanth972 @wongma7 Hello, is there any review update for this PR?

**Comment by jsafrane** [comment]
@Ashley-wenyizha it has been almost half a year. And it can significantly reduce memory footprint when a large GID range is set in StorageClass, from GB (!) to few MB and more frequent ListAccessPoint calls.

**Comment by Ashley-wenyizha** [comment]
hi @jsafrane 

Sincere apologize that this has been delayed as we were working on Access point increase earlier this year and the performance of listing all APs for customer has up to 1k AP might introduce performance issue.

We do benefit from reducing memory footprint but we need to give a test on the perf as well.

We have raised this to management attention and will grab more resource/bandwidth to work on this.

Meanwhile could you give it a fix on the conflicts? 

thanks!

**Comment by RomanBednar** [comment]
@Ashley-wenyizha Thank you for the update, PR should be ok now. Can you get back to us when you have an update on on the estimated completion for performance tests?

**Comment by Ashley-wenyizha** [comment]
Hi @RomanBednar 

Apologize for the late reply, has been oncall until this week. We already assigned engineer to look at this, will give you the test result update soon once I hear back within this next week.

Thanks for your patience again.

**Comment by Ashley-wenyizha** [comment]
Hi @RomanBednar 

Testing looks good, if not hearing back we will fix the merge conflict and merge in early next week.

Thanks for your contribution!

**Comment by RomanBednar** [comment]
@Ashley-wenyizha Thank you for the update, I've resolved the conflicts now.

**Comment by johnpmayer** [comment]
Is the plan to leave `ACCESS_POINT_PER_FS_LIMIT` at the current value (`120`)?

I see this comment marked resolved - https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/850#discussion_r1193520702

I'd definitely like to increase the value in my deployment. Is there still a performance concern listing large numbers of access points? Can we make this value configurable, so that users can make the trade-off themselves?

**Comment by mskanth972** [comment]
@johnpmayer, No the comment is just telling that Access points limit has been increased to 1000 from 120 recently.
https://www.amazonaws.cn/en/new/2023/amazon-efs-increases-the-maximum-number-of-access-points-per-file-system/

**Comment by johnpmayer** [comment]
I see this is being address in https://github.com/kubernetes-sigs/aws-efs-csi-driver/pull/1119
